### PR TITLE
Add option -m, --minimize to build mini/lite image [REVPI-2413]

### PR DESCRIPTION
**Pull request by zhan-kunbus**
Signed-off-by: Zhi Han <z.han@kunbus.com>

### CODE REVIEW

**Review by iluminat23 — CHANGES_REQUESTED** [review]
Please add a proper commit message. Also the minimize option is not tight to the lite image. So the summary is misleading.
Also I don't get why you want to duplicate the debs to download and debs to remove files? This makes it really hard to maintain them.

**Comment by zhan-kunbus** [comment]
> Please add a proper commit message.  Also the minimize option is not tight to the lite image. So the summary is misleading. 

pls give a name for the image

>Also I don't get why you want to duplicate the debs to download and debs to remove files? This makes it really hard to maintain them.

The lists of debs are different between normal and lite image.

**Comment by iluminat23** [comment]
It doesn't concern any one if a patch is done to refactor code or does something more. You might put this in the message, but it wastes just space in the headline.
Also the white space patch changes files which are installed in the rfs. So this patch doesn't just refactor the code. It has changes to the resulting files. They might most likely not have any impact. But it is not just refactoring!

File: customize_image.sh [hunk 1]
```diff
@@ -85,7 +83,10 @@ cleanup() {
 	cleanup_losetup
 }
 
+# Enable the Exist when non-zero return get catched
+set -ex
```
Comment (iluminat23) [inline customize_image.sh:118]: I would if anything move it to the top of the file. Then it is clear that this is globally set. As this option is not only relevant for the trap.

File: customize_image.sh [hunk 2]
```diff
@@ -1,10 +1,12 @@
 #!/bin/bash
 # customize raspbian image for revolution pi
 
-if [ "$#" != 1 ] ; then
-	echo 1>&1 "Usage: `basename "$0"` <image>"
-	exit 1
-fi
+usage () {
+	echo 'usage: customize_image.sh [-m, --minimize | -h, --help] <image>
+  -m, --minimize	build the minimized image
```
Comment (iluminat23) [inline customize_image.sh:6]: Add an explanation what this means. The current help is as useful as: Glass is transparent because it is made out of glass.

File: customize_image.sh [hunk 3]
```diff
@@ -37,6 +39,35 @@ if [ $$ != 2 ] && [ -x /usr/bin/newpid ] ; then
...
+		-m|--minimize) MINIMG=1 ; shift ;;
+		-h|--help) usage ; exit 0;;
+		*) shift; break ;;
+	esac
+done
+
+if [ "$MINIMG" != "1" ]; then
+	echo "The normal image will be built."
+else
+	echo "The lite image will be built."
```
[the remark this answers is not in the document]
Comment (iluminat23) [reply customize_image.sh:62]: And still the minimal image is not the lite image!
And the term normal image is also somewhat strange.
Why not just print a message "Building minimal image" and just don't print anything if the option is missing?

File: customize_image.sh [hunk 4]
```diff
@@ -37,6 +39,35 @@ if [ $$ != 2 ] && [ -x /usr/bin/newpid ] ; then
...
+fi
+
+if [ "$#" != 1 ] ; then
+	echo 1>&1 "Error: Cann't get path of <image> from parameter"
```
Comment (iluminat23) [inline customize_image.sh:66]: Cann't?

Also this seems like a different change, so do a separate commit for this.

**Comment by iluminat23** [comment]
The is also no documentation. Please update the README and extend it to document the new behavior.

**Review by iluminat23 — CHANGES_REQUESTED** [review]

**Comment by zhan-kunbus** [comment]
> Also the white space patch changes files which are installed in the rfs. So this patch doesn't just refactor the code. It has changes to the resulting files. They might most likely not have any impact. But it is not just refactoring!

The Tab will not be rendered to file by cat here, so the file in rfs will have no difference.

File: customize_image.sh (hunk 1 above)
Comment (zhan-kunbus) [reply customize_image.sh:118]: with -x the output message (usage() or error ) will be contaminated by the executed commands
 _-x  Print commands and their arguments as they are executed._

Comment (zhan-kunbus) [reply customize_image.sh:118]: the -e cause exist when command ends with non-zero, which is also not wished.
_-e  Exit immediately if a command exits with a non-zero status._

Comment (iluminat23) [reply customize_image.sh:118]: exist != exit
Yes -x is completely independent from the trap. So putting this set command above the trap is misleading. If you want to move it, move it to the top of the file. This is the place where this is expected.

File: customize_image.sh [hunk 5]
```diff
@@ -255,24 +292,27 @@ if [ -e "$IMAGEDIR/etc/init.d/apache2" ] ; then
...
 chroot "$IMAGEDIR" systemctl disable hciuart
 
 # disable 3rd party software
-chroot "$IMAGEDIR" systemctl disable logiclab
-chroot "$IMAGEDIR" systemctl disable nodered
+if [ "$MINIMG" != "1" ]; then
+	chroot "$IMAGEDIR" systemctl disable logiclab
+	chroot "$IMAGEDIR" systemctl disable nodered
+fi
 chroot "$IMAGEDIR" systemctl disable noderedrevpinodes-server
```
Comment (ferdrich-kunbus) [inline customize_image.sh:316]: Would it make sense to group all additions for a "full" image into one if instead of cluttering it all over the file? Or are I'm missing some dependencies?

```
chroot "$IMAGEDIR" apt-get -y install `egrep -v '^#' "$BAKERYDIR/min-debs-to-download"`
dpkg --root "$IMAGEDIR" --force-depends --purge rpd-wallpaper
chroot "$IMAGEDIR" apt-get -y install revpi-wallpaper
chroot "$IMAGEDIR" apt-get update

if [ "$MINIMG" != "1" ]; then
	chroot "$IMAGEDIR" apt-get -y install `egrep -v '^#' "$BAKERYDIR/debs-to-download"`
	chroot "$IMAGEDIR" apt-get -y install teamviewer-revpi
	
	# install nodejs and nodered with an install script and revpi-nodes from npm repository
	NODEREDSCRIPT="/tmp/update-nodejs-and-nodered.sh"
	/usr/bin/curl -sL \
		https://raw.githubusercontent.com/node-red/linux-installers/master/deb/update-nodejs-and-nodered\
		--output "$IMAGEDIR/$NODEREDSCRIPT"
	chmod 755 "$IMAGEDIR/$NODEREDSCRIPT"
	chroot "$IMAGEDIR" /usr/bin/sudo -u pi $NODEREDSCRIPT --confirm-install --confirm-pi
	rm "$IMAGEDIR/$NODEREDSCRIPT"
	chroot "$IMAGEDIR" /usr/bin/sudo -u pi /usr/bin/npm install --prefix /home/pi/.node-red node-red-contrib-revpi-nodes

	chroot "$IMAGEDIR" systemctl disable logiclab
	chroot "$IMAGEDIR" systemctl disable nodered
fi

# enable ssh daemon by default, disable swap, disable bluetooth on mini-uart
chroot "$IMAGEDIR" systemctl enable ssh
chroot "$IMAGEDIR" systemctl disable dphys-swapfile
chroot "$IMAGEDIR" systemctl disable hciuart
```

File: customize_image.sh (hunk 1 above)
Comment (zhan-kunbus) [reply customize_image.sh:118]: thanks for the catch of "exist => exit". 
And for set -e, would take another example of following code, if the set -x is used before it, the "if ... " will not be executed, but I wish it get executed to give a usage message.
```
# get the options
MYOPTS=`getopt -o mh --long minimize,help -- "$@"`
if [ "$?" != 0 ]; then  usage; exit 1; fi
```

Comment (zhan-kunbus) [reply customize_image.sh:118]: And for the set -x, with it on the top, the output of 
`~/revpi/imagebakery$ ./customize_image.sh --noexistoptions
getopt: unrecognized option '--noexistoptions'
`
would be 
`~/revpi/imagebakery$ ./customize_image.sh --noexistoptions
++ which curl
+ '[' '!' -x /usr/bin/curl ']'
++ which fsck.vfat
+ '[' '!' -x /usr/sbin/fsck.vfat ']'
++ which lsof
+ '[' '!' -x /usr/bin/lsof ']'
++ which parted
+ PARTED=/usr/sbin/parted
+ '[' x/usr/sbin/parted = x ']'
+ '[' '!' -x /usr/sbin/parted ']'
+ '[' 1036 '!=' 2 ']'
+ '[' -x /usr/bin/newpid ']'
+ MINIMG=0
++ getopt -o mh --long minimize,help -- --noexistoptions
getopt: unrecognized option '--noexistoptions'
+ MYOPTS=' --'
+ '[' 1 '!=' 0 ']'
+ usage
+ echo 'usage: customize_image.sh [-m, --minimize | -h, --help] <image>
  -m, --minimize        build the minimized image
  -h, --help            print the usage page
If no option is specified the normal image will be built.'
usage: customize_image.sh [-m, --minimize | -h, --help] <image>
  -m, --minimize        build the minimized image
  -h, --help            print the usage page
If no option is specified the normal image will be built.
+ exit 1
`

Comment (ferdrich-kunbus) [reply customize_image.sh:118]: Personally I use -x only for debugging purpose, not for script that will be delivered.

File: customize_image.sh (hunk 5 above)
Comment (zhan-kunbus) [reply customize_image.sh:316]: yes, as I remember, there are some dependencies between them.
Would think they can be removed somehow, however, it seems is another topic.

File: customize_image.sh (hunk 2 above)
Comment (zhan-kunbus) [reply customize_image.sh:6]: could you give an explanation?

File: customize_image.sh (hunk 4 above)
Comment (zhan-kunbus) [reply customize_image.sh:66]: Cann't => Can't thanks for the catch.

This was on the top to check the number of parameters to see whether path of image available (although it not a strict check). when check failed, the usage mesage would be given.   In this commit, as the number of parameters get manipulated by "getopt and set --", this check should be placed after that.  And it is not proper to just print the usage message, so I add the reason to it, and use the new added usage() function to show the usage message.   I see that they are somewhat integrated and therefor let them in one commit.

File: customize_image.sh (hunk 1 above)
Comment (zhan-kunbus) [reply customize_image.sh:116]: good idea, I will try it without -x

Comment (iluminat23) [reply customize_image.sh]: Please leave the -x option for now. It is not part of your ticket to change the overall output of the tool. We have really bad error handling in the imagebakery. Removing the -x option will leave us with no clue what went wrong if a customer reports an error.

File: customize_image.sh [hunk 6]
```diff
@@ -84,7 +84,7 @@ cleanup() {
 }
 
 # Enable the exit when non-zero return get catched
-set -ex
+set -e
```
Comment (iluminat23) [inline customize_image.sh:116]: NO

**Review by iluminat23 — CHANGES_REQUESTED** [review]

File: customize_image.sh (hunk 1 above)
Comment (iluminat23) [reply customize_image.sh]: No, not now. Your ticket is clear, please do what the ticket describes.

File: customize_image.sh (hunk 2 above)
Comment (zhan-kunbus) [reply customize_image.sh:6]: Should those information be kept in Release Node?   It almost the description of difference between "normal" image and "minimal" image, would be relative long and be changed possibly for every Release.

File: customize_image.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhan-kunbus) [reply customize_image.sh]: For the better communication,  within us and between us and users.  IMHO, we should name the things we generate, and a beautiful name might not needed, but it should be unambiguous

File: customize_image.sh [hunk 7]
```diff
@@ -196,7 +228,7 @@ sed -i -e '6i# allow-hotplug eth0\n# allow-hotplug eth1\n' "$IMAGEDIR/etc/networ
 # provide WPA template and prioritize wlan0 routes by default
 sed -i -e '/country=GB/d' "$IMAGEDIR/etc/wpa_supplicant/wpa_supplicant.conf"
 cat >> "$IMAGEDIR/etc/wpa_supplicant/wpa_supplicant.conf" <<-EOF
-
```
[the remark this answers is not in the document]
Comment (zhan-kunbus) [reply customize_image.sh:199]: OK, I don't know how to describe the reason, I will remove this commit and change the setting of my VS Code to make it compatible with this kind of issue.

File: customize_image.sh (hunk 5 above)
Comment (nbuchwitz) [reply customize_image.sh:315]: @ferdrich-kunbus I like your idea of moving most of the additional parts into one group, but would suggest to address this in a separate PR

**Review by nbuchwitz — APPROVED** [review]

**Review by ferdrich-kunbus — APPROVED** [review]
Looks ok so far

**Review by iluminat23 — APPROVED** [review]
Looks good
